src/Database/Esqueleto/Experimental.hs

```diff
@@ -779,103 +750,149 @@ type family ErrorOnLateral a :: Constraint where
 
 {-- Type class magic to allow the use of the `InnerJoin` family of data constructors in from --}
 class ToFrom a where
+    type ToFromT a
     toFrom :: a -> From (ToFromT a)
 
-instance ToFrom (From a) where
-    toFrom = id
+type family FromOnClause a where
+    FromOnClause (a, b -> SqlExpr (Value Bool)) = b
+    FromOnClause a = TypeError ('Text "Missing ON clause")
 
 instance {-# OVERLAPPABLE #-} ToFrom (InnerJoin a b) where
+    type ToFromT (InnerJoin a b) = FromOnClause b
     toFrom = undefined
 instance {-# OVERLAPPABLE #-} ToFrom (LeftOuterJoin a b) where
+    type ToFromT (LeftOuterJoin a b) = FromOnClause b
     toFrom = undefined
-instance {-# OVERLAPPABLE #-} ToFrom (RightOuterJoin a b) where
-   toFrom = undefined
 instance {-# OVERLAPPABLE #-} ToFrom (FullOuterJoin a b) where
+    type ToFromT (FullOuterJoin a b) = FromOnClause b
+    toFrom = undefined
+instance {-# OVERLAPPABLE #-} ToFrom (RightOuterJoin a b) where
+    type ToFromT (RightOuterJoin a b) = FromOnClause b
+    toFrom = undefined
```

Comment:
These `undefined`s are kinda weirding me out. If we're not ever calling them, should we delete the method? If we are calling them, shouldn't we implement these?

Reply:
These instances are the defaults that throw the type error. They are undefined to suppress the warning caused by not defining ToFrom.

Reply:
If you have a better implementation I am not stoked about how the Type errors work. But it's still better than what was there before in my mind

Reply:
Ooh, I see. It's a bit indirected, but it makes sense.